### PR TITLE
h2/fix dropper

### PR DESCRIPTION
**Pull request by howardjohn**
- **With compiler bug**
- **h2: remove drop counter from server**

### CODE REVIEW

File: src/proxy/h2.rs [hunk 1]
```diff
@@ -106,36 +126,16 @@ impl H2StreamWriteHalf {
...
-        }
-    }
-}
-
-impl Drop for H2StreamWriteHalf {
+impl Drop for DropCounter {
     fn drop(&mut self) {
         let mut half_dropped = Arc::new(());
         std::mem::swap(&mut self.half_dropped, &mut half_dropped);
         if Arc::into_inner(half_dropped).is_none() {
```
Comment (ilrudie) [inline src/proxy/h2.rs:133]: I know you aren't changing this but it seems like this logic isn't correct... If we get `none` back it means we are **not** the last strong reference so the other half isn't dropped. If we get the inner value back we are the last strong reference.

File: src/proxy/h2.rs [hunk 2]
```diff
@@ -71,21 +71,41 @@ async fn do_ping_pong(
...
-    pub half_dropped: Arc<()>,
-    pub active_count: Arc<AtomicU16>,
+    send_stream: h2::SendStream<SendBuf>,
+    _dropped: Option<DropCounter>,
+}
+
+struct DropCounter {
+    // Whether the other end of this shared counter has already dropped.
+    // We only decrement if they have, so we do not double count
+    half_dropped: Arc<()>,
```
Comment (bleggett) [inline src/proxy/h2.rs:92]: Could we just use an AtomicBool here?

Comment (howardjohn) [reply src/proxy/h2.rs:92]: Yeah this is basically the same thing but it makes me feel smart :slightly_smiling_face:

Comment (howardjohn) [reply src/proxy/h2.rs:92]: I think the reason I didn't do this was it would need to be Arc<AtomicBool>. Would work though

Comment (bleggett) [reply src/proxy/h2.rs:92]: If it's gotta be Arc'd anyway then nevermind yea

**Review by bleggett — APPROVED** [review]
Nit is that I'm pretty sure there's a simpler way to do the same thing without the 

`new + swap` stuff, but that's mostly cosmetic, the fix logic is good.

**Pull request merged**